### ISSUE TITLE
useInfiniteQuery doesn't invalidate cursors

### ISSUE THREAD

**Issue by KrasiNedew** [tannerlinsley/react-query#407]
#### First, big thanks for the great work you're doing! The libs you bring to the community are top-notch!

However I believe there's a fundamental flaw in the `useInfiniteQuery` hook implementation.
It stores `pageVariables` (I'll call them cursors) and on refetch tries to get data based on them in parallel. That works if the underlying data never changes. But this is rarely the case. Usually with cursor based pagination, cursors shouldn't be stored, because if the data is ever mutated, they should be invalidated.

#### I'll give a simple example here:
1. Let's assume that our cursor is just an id pointing to the first item on the next page.
2. Let's also assume that we have a limit of 3 results per page (for simplicity).
3. We're sending our first request and are getting back
```js
{ 
  nextCursor: 4, 
  results: [
   { id: 1 }, 
   { id: 2 }, 
   { id: 3 }
  ]
}
```
4. We scroll down a bit and it's time to fetch the next page, so we get back
```js
{ 
  nextCursor: 7, 
  results: [
   { id: 4 }, 
   { id: 5 }, 
   { id: 6 }
  ]
}
```
5. We scroll down even further
```js
{ 
  nextCursor: 10, 
  results: [
   { id: 7 }, 
   { id: 8 }, 
   { id: 9 }
  ]
}
```
6. At this point our data looks good
```js
[
  { 
    nextCursor: 4, 
    results: [
      { id: 1 }, 
      { id: 2 }, 
      { id: 3 }
    ]
  },
  { 
    nextCursor: 7, 
    results: [
      { id: 4 }, 
      { id: 5 }, 
      { id: 6 }
    ]
  },
  { 
    nextCursor: 10, 
    results: [
      { id: 7 }, 
      { id: 8 }, 
      { id: 9 }
    ]
  }
]
```
7. Our page variables (cursors) look like this
```js
[
  ['items'], // the first query doesn't have a cursor, since we don't know it yet
  ['items', 4], // the second query was executed with the cursor returned from the first one
  ['items', 7] // the third query was executed with the cursor returned from the second one
]
```
8. Let's remove the item with id = 5
```js
const [removeItem] = useMutation(deleteItem, {
  onSuccess: () => { queryCache.refetchQueries('items'); }
});

// in some event handler
removeItem(5);
```
9. When the item is deleted successfully, a refetch is triggered. I'll remind you here, that the stored cursors at this point are:
```js
[
  ['items'],
  ['items', 4],
  ['items', 7]
]
```
This gives us some duplicated results back (item 7), since the cursor for the 3rd query was invalidated (it should now be 8, but we're already storing 7 and executing the query with that)
```js
[
  { 
    nextCursor: 4, 
    results: [
      { id: 1 }, 
      { id: 2 }, 
      { id: 3 }
    ]
  },
  {
    nextCursor: 8, // the nextCursor is now 8, instead of 7, because we deleted item 5
    results: [
      { id: 4 }, 
      { id: 6 }, 
      { id: 7 }
    ]
  },
  // we're storing 7 and executing the 3rd query with it
  // even though the nextCursor from the previous query is now 8
  // our first result is a duplicate of the last result of the previous query
  {
    nextCursor: 10, 
    results: [
      { id: 7 }, 
      { id: 8 }, 
      { id: 9 }
    ]
  }
]
```
The same problem exists when we add new records and we use a descending sort, so that items are added at the beginning.

#### Proposed solution (might also be flawed)
We can't assume that the data mutation would come from our end and have some imperative handling, because some other user might mutate the data and it might change underneath us, even without us performing any action.

So instead of storing cursors and refetching in parallel, make the refetch sequential and call each query with a fresh cursor (call getFetchMore before each query).

PS: This solution might also have a flaw though - what if we insert an item at the beginning, which would offset the subsequent cursors by 1? I believe in this case we're losing the last record in our data set, which also isn't ideal...

**Comment by KrasiNedew** [comment]
I also want to mention that the flaw of potentially offsetting the cursors if an item is inserted at the beginning isn't that big of a deal, since it will be loaded with the next page anyway. The proposed solution eliminates the possibility of duplicates, which is a far bigger problem.

**Comment by tannerlinsley** [comment]
Yep this should be fixed for sure. Cursors indeed should never be stored and this use case is bulletproof.

**Comment by tannerlinsley** [comment]
Closed by #408 and #409
